certification/internal/policy/operator/has_v2_related_image_schema_version.go

```diff
@@ -0,0 +1,182 @@
+	relatedImages := csv.Spec.RelatedImages
+
+	// no related images == nothing to check
+	if len(relatedImages) == 0 {
+		return nil, nil
+	}
+
+	imageStrings := make([]string, len(relatedImages))
+	for i, relatedImage := range relatedImages {
+		imageStrings[i] = relatedImage.Image
+	}
+
+	return imageStrings, nil
+}
+
+func (p *RelatedImagesAreSchemaVersion2Check) inspectSchemaVersionForImage(images []string) (map[string]int, error) {
+	// TODO: handle cases where the related images are behind authentication.
+	imageToSchemaVersion := map[string]int{}
+	for _, image := range images {
+		manifestB, err := crane.Manifest(image)
+		if err != nil {
+			return nil, err
+		}
+
+		schemaVersion, err := p.getSchemaVersionFromRawManifest(manifestB)
+		if err != nil {
+			return nil, err
+		}
+
+		imageToSchemaVersion[image] = schemaVersion
+	}
+
+	return imageToSchemaVersion, nil
+}
+
+// getSchemaVersionFromRawManifest accepts the raw manifest blob, asserts that it is the expected format,
+// and returns the value of the top-level schemaVersion key.
+func (p *RelatedImagesAreSchemaVersion2Check) getSchemaVersionFromRawManifest(manifest []byte) (int, error) {
+	var rawManifest map[string]interface{}
+	if err := json.Unmarshal(manifest, &rawManifest); err != nil {
+		log.Error("unable to unmarshal inspected manifest")
+		return 0, err
+	}
+
+	schemaVersionIface, found := rawManifest["schemaVersion"]
+	if !found {
+		return 0, fmt.Errorf("rawManifest is an unexpected format")
+	}
+
+	// when unmarshalled by json, number will be a float64
+	schemaVersionf64, ok := schemaVersionIface.(float64)
+	if !ok {
+		return 0, fmt.Errorf("schemaVersion value is an unexpected type")
+	}
+
+	// We don't expect the schemaVersion value to be large, so converting to
+	// an int should be okay.
+	schemaVersion := int(schemaVersionf64)
+
+	return schemaVersion, nil
+}
+
+// validate checks each image's schema value and returns true if they're all version 2.
+func (p *RelatedImagesAreSchemaVersion2Check) validate(imageToSchemaVersionMap map[string]int) (bool, error) {
+
+	// if the map is empty or nil, then we auto pass this check.
+	// note that internally, len(nil) == 0 according to staticcheck
+	if len(imageToSchemaVersionMap) == 0 {
+		return true, nil
+	}
+
+	v2ImageManifests := []string{}
+	notv2ImageManifests := []string{}
+
+	for image, schemaVersion := range imageToSchemaVersionMap {
+		if schemaVersion == 2 {
+			v2ImageManifests = append(v2ImageManifests, image)
+		} else {
+			notv2ImageManifests = append(notv2ImageManifests, image)
+		}
+	}
+
+	// if we determine that all images are using schemaVersion 2, then we pass the check
+	if len(v2ImageManifests) == len(imageToSchemaVersionMap) {
+		return true, nil
+	}
+
+	log.Info("the following related images found in CSV were not image manifests with schema version 2", notv2ImageManifests)
+	return false, nil
```

Comment:
A nit: I would have the log (LoC 158) printed in an if block and use return len(notv2ImageManifests) == 0, nil as the return statement.

Reply:
I personally prefer the explicit return, as I believe it makes it a bit easier to follow in debug cases.

If I return the `len` evaluation, then a human debugger reviewing this would need to evaluate if there are any places in the code where the input to `len()` would have changed from what they expected. I feel it makes things a bit more cumbersome to troubleshoot.